Client-supplied cache keys (Spice-Cache-Key)

mach-kernel opened the pull request:
## 📝 Summary

Users may provide an `X-Spice-Cache-Key` header value as an alternative to plan and SQL string based caching. There are no changes to developer UX w.r.t. configuration. If a valid** cache key is provided, then it is used for result caching. Invalid (or not-provided) keys fall back to the spicepod's configured behavior.

## TODOs
- [x] Bind user cache key in necessary contexts
- [x] Sanitize cache key + warn if invalid
- [x] Implement for `datafusion::query::cache` (...`/v1/sql`)
- [x] Implement for `/v1/search`
- [x] `Vary` header
- [x] Update docs
  - [x] Basic narrative
  - [x] Docs repo PR
- [x] Add to `e2e_test_ci`

## 🔗 Related

Resolves #6413 

## 👀 Notes for Reviewers

- New `CacheKey::User` variant
- New `CacheResponse` struct (feel free to suggest a better name) to encapsulate commonly used context in `datafusion::query::cache`
  - DRY up some logic around e.g. `try_get_cached_result` handling
  - Simplifies `CacheResult` enum (seems we always want to have a `QueryTracker` on hand)
- Allows cache keys with alphanumeric values (plus `'_'` `'-'` up to 128 chars (supports bigger hashes e.g. sha512)
- Add `Vary: X-Spice-Cache-Key` to headers
  - ...there may be an existing `Vary` header in the response (i.e. you may see two lines with a Vary response). This is [allowed for headers that have list values](https://stackoverflow.com/a/60060332)

## 📚 Docs

https://github.com/spiceai/docs/pull/1076

mach-kernel committed:
[1/?] thread a user cache key into context harness

mach-kernel committed:
[2/3] add cache key sanitization with warning for invalid keys, test for binding user key to context

mach-kernel committed:
[3/3] write tests for datafusion/user key

mach-kernel committed:
X-Spice-Cache-Key + /v1/search

mach-kernel committed:
decorate vary header if user cache key is set

mach-kernel committed:
lint

lukekim commented on crates/runtime/src/request/cache_control.rs:
```diff
@@ -20,13 +20,15 @@ use app::App;
     /// Use the raw input (e.g., unparsed SQL string) as the cache key.
     Raw,
+    /// Use a user-provided cache key via the X-Spice-Cache-Key header.
+    User,
```
Let's call this `Client` or `ClientSupplied` as our "users" are generally machines (apps/services/agents).

mach-kernel replied:
OK -- I like `ClientSupplied`!

lukekim commented on crates/runtime/src/request/context.rs:
```diff
@@ -329,20 +357,80 @@ impl RequestContextBuilder {
     }
+
+    fn sanitize_cache_key(key: String) -> Option<String> {
+        if USER_CACHE_KEY_REGEX.is_match(&key) {
```
Will regex in the critical query path meaningfully affect performance?

mach-kernel replied:
No, the regexp is compiled once and the expression itself has anchors + match limits

mach-kernel committed:
CacheKey::User -> CacheKey::ClientSupplied

mach-kernel committed:
Merge branch 'trunk' into mach/x-spice-cache-key

phillipleblanc commented on crates/runtime/src/datafusion/query/cache.rs:
```diff
@@ -166,33 +197,39 @@ impl Query {
             }
             _ => {
-                return Ok((CacheResult::WrongCacheKeyType(tracker), None));
+                return Ok(CacheResponse::from(
```
You will also need to pass the query tracker here too right?

mach-kernel replied:
Yes, good catch

phillipleblanc commented on crates/runtime/src/http/v1/mod.rs:
```diff
@@ -262,6 +271,11 @@ fn attach_cache_headers(headers: &mut HeaderMap, results_cache_status: CacheStat
+
+    // Tell CDN entry is unique per user cache key
+    if user_key_specified {
+        headers.insert("Vary", HeaderValue::from_static("X-Spice-Cache-Key"));
```
```suggestion
        headers.insert("Vary", HeaderValue::from_static("Spice-Cache-Key"));
```

https://www.rfc-editor.org/rfc/rfc6648

phillipleblanc commented on crates/runtime/src/http/v1/search.rs:
```diff
@@ -163,6 +167,11 @@ pub(crate) async fn post(
 
+                // Tell CDN entry is unique per user cache key
+                if request_context.user_cache_key().is_some() {
+                    headers.insert("Vary", HeaderValue::from_static("X-Spice-Cache-Key"));
```
```suggestion
                    headers.insert("Vary", HeaderValue::from_static("Spice-Cache-Key"));
```

phillipleblanc commented on crates/runtime/src/request/cache_control.rs:
```diff
@@ -72,17 +74,13 @@ impl CacheControl {
-        };
-        let Ok(cache_control_str) = cache_control.to_str() else {
-            return Self::Cache(cache_key_type);
+        let cache_key_type = match headers.get("x-spice-cache-key") {
```
```suggestion
        let cache_key_type = match headers.get("spice-cache-key") {
```

phillipleblanc commented on crates/runtime/src/request/context.rs:
```diff
@@ -257,6 +271,14 @@ impl RequestContextBuilder {
         self.cache_control = CacheControl::from_headers(headers);
+        self.user_cache_key = match self.cache_control {
+            CacheControl::Cache(CacheKeyType::ClientSupplied) => headers
+                .get("x-spice-cache-key")
```
```suggestion
                .get("spice-cache-key")
```

phillipleblanc commented on crates/runtime/src/request/context.rs:
```diff
@@ -329,20 +357,80 @@ impl RequestContextBuilder {
+            Some(key)
+        } else {
+            warn!(
+                "X-Spice-Cache-Key provided for request ({}) is invalid. A valid cache key is \
```
```suggestion
                "Spice-Cache-Key provided for request ({}) is invalid. A valid cache key is \
```

phillipleblanc commented on crates/runtime/src/request/context.rs:
```diff
@@ -329,20 +357,80 @@ impl RequestContextBuilder {
+        if USER_CACHE_KEY_REGEX.is_match(&key) {
+            Some(key)
+        } else {
+            warn!(
```
I don't think its a good idea to warn here - we could add this to the task history, but this gives a potential attacker an avenue to spam the logs, which could be used to cover up other malicious activity or make it harder to understand what Spice is doing.

phillipleblanc commented on crates/runtime/src/request/context.rs:
```diff
@@ -329,20 +357,80 @@ impl RequestContextBuilder {
+        headers.append("cache-control", HeaderValue::from_static("cache"));
+
+        // Test user-provided cache key
+        headers.append("x-spice-cache-key", HeaderValue::from_static("foo"));
```
```suggestion
        headers.append("spice-cache-key", HeaderValue::from_static("foo"));
```

phillipleblanc commented on crates/runtime/src/request/context.rs:
```diff
@@ -329,20 +357,80 @@ impl RequestContextBuilder {
+
+        // Test invalid user cache key falling back to default behavior
+        headers.remove("x-spice-cache-key");
+        headers.append("x-spice-cache-key", HeaderValue::from_static("foo$$"));
```
```suggestion
        headers.remove("spice-cache-key");
        headers.append("spice-cache-key", HeaderValue::from_static("foo$$"));
```

Jeadie commented on crates/runtime/src/request/context.rs:
```diff
@@ -51,6 +54,10 @@ tokio::task_local! {
 
+#[allow(clippy::expect_used)]
+static USER_CACHE_KEY_REGEX: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^([\w-]{1,128})$").expect("Must compile regex"));
```
nit: we generally would use `unreachable!` for something like this.

mach-kernel replied:
TIL, thanks!

phillipleblanc committed:
Merge branch 'trunk' into mach/x-spice-cache-key

mach-kernel committed:
feedback

- x-spice-cache-key -> spice-cache-key
- missing query tracker for codepath
- unreachable! for regex compilation
- do not warn

mach-kernel committed:
Merge branch 'trunk' into mach/x-spice-cache-key

mach-kernel committed:
- bug: ensure request cache key is bound to the correct corresponding key for the cache control setting
- misc: rename remainder of "user_cache_key" -> "client_supplied_cache_key"

mach-kernel committed:
add client cache key tests to e2e_test_ci

sgrebnov commented on crates/runtime/src/datafusion/query/cache.rs:
```diff
@@ -80,28 +105,34 @@ impl Query {
-        let (tracker, cache_status, sql_raw_cache_key) = match Self::try_get_cached_result(
+        let sql_or_user_cache_key = match request_context.client_supplied_cache_key() {
+            Some(user_key) => CacheKey::ClientSupplied(user_key),
+            _ => sql_cache_key,
```
Should we move `CacheKey::Query(sql, parameters.as_ref());` from line 108 here?

sgrebnov replied:
I see - there is special logic where  sql_cache_key is required: `// Always use CacheKey::Query when checking the plan cache`
